docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,960 @@
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix is the primary mechanism for protecting against split-brain situations, and
+enabling safe multi-attachment of tenants:
+
+- Two pageservers running with the same node ID (e.g. after a failure, where there is
+  some rogue pageserver still running) will not try to write to the same objects.
+- Multiple attachments of the same tenant will not try to write to the same objects, as
+  each attachment would have a distinct attachment generation.
+
+### Deletion Part 1: Remote object deletions
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. it is the legitimate holder of its node ID), _and_ that its attachments for the tenants
+  whose data is being deleted are current.
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation for the node and the attachment(s) doing deletion.
+4. Actually delete the objects
+
+Note that at stage 3 we are only confirming that deletions of objects no longer referenced
+by the metadata written in step 2 are safe. While Step 3 is in flight, new deletions may
+be enqueued, but these are not safe to execute until a subsequent iteration of the
+generation check.
+
+Because step 3 puts load on the control plane, deletion lists should be executed lazily,
+accumulating many deletions before executing them behind a batched generation check message
+to the control plane, which would contain a list of tenant generation numbers to validate.
+
+The result of the request to the control plane may indicate that only some of the attachment
+generations are fresh: if this is the case, then the pageserver must selectively drop the
+deletions from a stale generation, but still execute the deletions for attachments
+that had a fresh generation.
```

Comment:
How can it happen that just some attachment generations are fresh?
How does the batch even look like?

I think it would be helpful to repeat the "order of operations for correctness" above with the batching.

Reply:
I've shortened this section in 6dbe92ad51d9c926ddd2fa3ee7b73891b187eb3f and pushed queue handling detail to the optimizations section, with some a more explicit format spec.